### ISSUE TITLE
will this work with direct logic series?

### ISSUE THREAD

**Issue by nagimov** [numat/clickplc#2]
Not a PLC guy so please bear with me :)
I would like to add support for [direct logic series 405](https://www.automationdirect.com/adc/overview/catalog/programmable_controllers/directlogic_series_plcs_(micro_to_small,_brick_-a-_modular)/directlogic_405_(small_modular_plc)) PLC. Our system is up and running and I can't test things right away, so I'm a bit stuck in theoretical research stage right now.
From what I understand (please correct me if I'm wrong), the only specificity of the clickPLC here is mapping of PLC memory ranges (X, Y, etc) to modbus addresses, and it should be (relatively) easy to remap for DL405.
There is a nice table of modbus memory addressing for DL405 in the [manual](https://cdn.automationdirect.com/static/manuals/d4454userm/d4454m.pdf) (chapter 4-23) with some useful examples, description of data types, etc.:
<img src="https://user-images.githubusercontent.com/3197765/54955630-13fb9480-4f0b-11e9-9571-81f65eeb6afd.png" width=50%>
I however haven't found same info for clickPLC, would you mind pointing me out?
Could you point at any other quirks I need to know?
Thanks!!

**Comment by patrickfuller** [comment]
Happy to help!

This project's only a few hundred lines of code, but quirks with PLCs (or anything industrial) are everywhere. Some off the top of my head:

 * MODBUS addresses can start counting at 0 or 1, depending on what the equipment developer felt like at the time
 * MODBUS has a max packet payload of 104 bytes (52 registers), although some equipment cuts out at other spots
 * MODBUS doesn't support multiple clients or asynchronous requests. It will usually silently drop all received packets if it's busy.
 * The registers and coils don't have to be in sequential addresses. The ClickPLC's coils jump 32 bits forward from time to time.
 * Registers = 16bits, but store floats. They don't always implement an IEEE float properly and numbers can behave weirdly. You'll get rounding 2-3 decimals out.
 * Adding registers adds an extra layer of complexity to little vs big endian. You can be little endian in bits but big endian in registers (e.g. Aaaaaaaa Bbbbbbbb vs. Bbbbbbbb Aaaaaaaa). What's used is whatever the equipment developer felt like at the time, so you have to guess and check all four combinations

Basically, industrial controls are terrible. If a PLC properly supported e.g. HTTP REST, you'd put a lot of control engineers out of a job.

My recommendation for you would be to fork this repo, read the code, and try copying the pattern. You'll have to guess and check and you'll find lots of quirks along the way. Make sure you document them as you go and you should be able to get to something functional.

Hope this helps!

**Comment by nagimov** [comment]
Yeah thanks this project is by far the most comprehendable and well documented one dealing with modbus comm I have so far discovered. Yet still it packs as much demystification of modbus as possible (I guess it's written to be actually used :+1: )
Another feature I spend two nights crying over is DL405's amazing engineering solution of storing decimal 1234 as HEX 0x1234. Pure genius.
Thanks a bunch for tips! We do apparently have a spare PLC so I don't have an excuse to not to work on that anymore.
